mayastor/src/bdev/nexus/nexus_metadata.rs

```diff
@@ -5,821 +5,745 @@
+    ReadAlloc { source: DmaError, name: String },
     #[snafu(display(
         "Failed to allocate buffer for writing {}: {}",
         name,
         source
     ))]
-    WriteAlloc { name: String, source: DmaError },
-    #[snafu(display("Serialization error: {}", source))]
-    SerializeError { source: Error },
-    #[snafu(display("Deserialization error: {}", source))]
-    DeserializeError { source: Error },
+    WriteAlloc { source: DmaError, name: String },
+    #[snafu(display("Error reading {}: {}", name, source))]
+    ReadError { source: CoreError, name: String },
+    #[snafu(display("Error writing {}: {}", name, source))]
+    WriteError { source: CoreError, name: String },
     #[snafu(display(
-        "Incorrect MetaData header size: actual={} expected={}",
-        actual_size,
-        expected_size
+        "Failed to obtain BdevHandle for child {}: {}",
+        name,
+        source
     ))]
-    HeaderSize {
-        actual_size: u32,
-        expected_size: u32,
-    },
-    #[snafu(display("Incorrect MetaData header signature"))]
-    HeaderSignature {},
-    #[snafu(display("Incorrect MetaData header checksum"))]
-    HeaderChecksum {},
-    #[snafu(display("Incorrect MetaData index checksum"))]
-    IndexChecksum {},
-    #[snafu(display("Incorrect MetaData configuration object checksum"))]
+    HandleError { source: CoreError, name: String },
+    #[snafu(display("Incorrect MetaDataObject signature"))]
+    ObjectSignature {},
+    #[snafu(display("Incorrect MetaDataObject checksum"))]
     ObjectChecksum {},
-    #[snafu(display("MetaData index is inconsistent"))]
-    IndexInconsistent {},
-    #[snafu(display("MetaData index ({}) out of range ({})", selected, used))]
-    IndexOutOfRange { selected: u32, used: u32 },
-    #[snafu(display(
-        "Number of objects ({}) exceeds index size ({})",
-        used,
-        size
-    ))]
-    IndexSizeExceeded { used: u32, size: u32 },
-    #[snafu(display("Not enough space left on MetaData partition"))]
-    PartitionSizeExceeded {},
-    #[snafu(display("MetaData partition is missing or invalid"))]
-    MissingPartition {},
-    #[snafu(display("Error calculating timestamp: {}", source))]
-    TimeStampError { source: SystemTimeError },
+    #[snafu(display("Incorrect child table checksum"))]
+    ChildTableChecksum {},
+    #[snafu(display("Maximum number of children exceeded"))]
+    ChildTableSize {},
+    #[snafu(display("Incorrect MetaDataIndex signature"))]
+    IndexSignature {},
+    #[snafu(display("Incorrect MetaDataIndex size"))]
+    IndexSize {},
+    #[snafu(display("Incorrect MetaDataIndex checksum"))]
+    IndexChecksum {},
+}
+
+#[derive(Debug, Default, Deserialize, PartialEq, Serialize)]
+pub struct MetaDataChildEntry {
+    pub guid: Guid,
+    pub state: u16,
+}
+
+impl MetaDataChildEntry {
+    /// Calculate checksum over the child entries
+    pub fn checksum(children: &[MetaDataChildEntry]) -> u32 {
```

Comment:
It does appear this function has some reasonable error conditions due to serialization. Are all of these situations ones in which we should panic?

Reply:
As far as I can tell, most of the variants of the relevant error type pertain to deserialization rather than serialization. The only possible errors here would result from an I/O error on the internal writer (Cursor) which, shouldn't ever fail as long as we don't run out of memory. But I take your point, it is still good practice to test for all (theoretically) possible error conditions.

Reply:
Have updated to return a Result rather than trigger a panic even though the error shouldn't really ever occur in practice